File: dashboard/android/app/src/main/java/com/attobot/dashboard/ui/screens/IndexesScreen.kt
package com.attobot.dashboard.ui.screens

import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import androidx.navigation.NavHostController
import com.attobot.dashboard.core.IndexRow
import com.attobot.dashboard.state.IndexesViewModel
import com.attobot.dashboard.state.UiState
import com.attobot.dashboard.ui.components.DataTable
import com.attobot.dashboard.ui.components.ErrorState
import com.attobot.dashboard.ui.components.FilterPills
import com.attobot.dashboard.ui.components.LoadingView
import com.attobot.dashboard.ui.components.PullRefreshScreen
import com.attobot.dashboard.ui.components.TableColumn
import com.attobot.dashboard.ui.components.TypePill
import com.attobot.dashboard.ui.theme.accent
import com.attobot.dashboard.ui.theme.muted
import com.attobot.dashboard.ui.theme.text

/**
 * Database indexes across all non-system schemas (attobot, df, …), including
 * pgvector's hnsw/ivfflat access methods. The schema filter is derived
 * client-side from the fetched rows (mirrors the web Indexes page).
 */
@Composable
fun IndexesScreen(@Suppress("UNUSED_PARAMETER") navController: NavHostController) {
    val vm: IndexesViewModel = viewModel()
    val state by vm.state.collectAsStateWithLifecycle()
    val schema by vm.schema.collectAsStateWithLifecycle()
    val s = state

    val all = (s as? UiState.Ready)?.data ?: emptyList()
    val schemas = all.map { it.schemaName }.distinct().sorted()
    val rows = if (schema.isBlank()) all else all.filter { it.schemaName == schema }

    PullRefreshScreen(refreshing = false, onRefresh = null) {
        Text(
            "Indexes",
            color = text,
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.padding(bottom = 12.dp),
        )
        FilterPills(options = schemas, value = schema, onSelect = vm::setSchema, allLabel = "all schemas")
        Text(
            "${rows.size} index${if (rows.size == 1) "" else "es"}",
            color = muted,
            fontSize = 12.sp,
            modifier = Modifier.padding(top = 4.dp, bottom = 12.dp),
        )
        when (s) {
            is UiState.Loading -> LoadingView()
            is UiState.Error -> ErrorState(s.message)
            is UiState.Ready -> DataTable(rows = rows, columns = indexColumns())
        }
    }
}

private fun indexColumns(): List<TableColumn<IndexRow>> = listOf(
    TableColumn(header = "schema", width = 90.dp) {
        Text(it.schemaName, color = text, fontFamily = FontFamily.Monospace, fontSize = 12.sp, maxLines = 1)
    },
    TableColumn(header = "table", width = 110.dp) {
        Text(it.tableName, color = text, fontFamily = FontFamily.Monospace, fontSize = 12.sp, maxLines = 1)
    },
    TableColumn(header = "index", width = 150.dp) {
        Text(it.indexName, color = accent, fontFamily = FontFamily.Monospace, fontSize = 12.sp, maxLines = 1)
    },
    TableColumn(header = "type", width = 80.dp) {
        TypePill(it.indexType)
    },
    TableColumn(header = "flags", width = 80.dp) {
        when {
            it.isPrimary -> TypePill("PK")
            it.isUnique -> TypePill("UNIQUE")
            else -> Text("—", color = muted, fontSize = 13.sp)
        }
    },
    TableColumn(header = "size", width = 90.dp) {
        Text(it.size, color = text, fontSize = 13.sp)
    },
    TableColumn(header = "scans", width = 90.dp) {
        Text(it.scans, color = text, fontSize = 13.sp)
    },
    TableColumn(header = "definition", width = 280.dp) {
        Text(it.definition, color = muted, fontFamily = FontFamily.Monospace, fontSize = 11.sp, maxLines = 2)
    },
)

File: dashboard/ios/OpenWorldDashboard/Views/Screens/IndexesScreen.swift
import SwiftUI

/// Database indexes across all non-system schemas (OpenWorld, df, …), including
/// pgvector's hnsw/ivfflat access methods once a vector column is indexed.
/// No poll; the schema filter is derived client-side from the fetched rows
/// (mirrors the web Indexes page).
struct IndexesScreen: View {
    @State private var res = Resource<[IndexRow]>()
    @State private var schema: String = ""

    private var all: [IndexRow] { res.value ?? [] }
    private var schemas: [String] { Array(Set(all.map { $0.schemaName })).sorted() }
    private var rows: [IndexRow] {
        schema.isEmpty ? all : all.filter { $0.schemaName == schema }
    }

    var body: some View {
        ScreenScroll {
            Text("Indexes")
                .font(.system(size: 20, weight: .bold))
                .foregroundStyle(Theme.text)
                .padding(.bottom, 12)

            FilterPills(options: schemas, value: schema, onSelect: { schema = $0 }, allLabel: "all schemas")
                .padding(.horizontal, -16)
                .padding(.bottom, 8)

            Text("\(rows.count) index\(rows.count == 1 ? "" : "es")")
                .font(.system(size: 12))
                .foregroundStyle(Theme.muted)
                .padding(.bottom, 12)

            if let err = res.error, res.value == nil {
                ErrorState(message: err)
            } else if res.isLoading && res.value == nil {
                Spinner()
            } else {
                DataTable(
                    columns: [
                        DataTableColumn<IndexRow>(id: "schema", title: "schema", width: 90) { r in
                            Text(r.schemaName).font(.system(size: 12, design: .monospaced)).foregroundStyle(Theme.text).lineLimit(1)
                        },
                        DataTableColumn<IndexRow>(id: "table", title: "table", width: 110) { r in
                            Text(r.tableName).font(.system(size: 12, design: .monospaced)).foregroundStyle(Theme.text).lineLimit(1)
                        },
                        DataTableColumn<IndexRow>(id: "index", title: "index", width: 150) { r in
                            Text(r.indexName).font(.system(size: 12, design: .monospaced)).foregroundStyle(Theme.accent).lineLimit(1)
                        },
                        DataTableColumn<IndexRow>(id: "type", title: "type", width: 80) { r in
                            TypePill(r.indexType)
                        },
                        DataTableColumn<IndexRow>(id: "flags", title: "flags", width: 80) { r in
                            if r.isPrimary {
                                TypePill("PK")
                            } else if r.isUnique {
                                TypePill("UNIQUE")
                            } else {
                                Text("—").font(.system(size: 13)).foregroundStyle(Theme.muted)
                            }
                        },
                        DataTableColumn<IndexRow>(id: "size", title: "size", width: 90) { r in
                            Text(r.size).font(.system(size: 13)).foregroundStyle(Theme.text)
                        },
                        DataTableColumn<IndexRow>(id: "scans", title: "scans", width: 90) { r in
                            Text(r.scans).font(.system(size: 13)).foregroundStyle(Theme.text)
                        },
                        DataTableColumn<IndexRow>(id: "def", title: "definition", width: 280) { r in
                            Text(r.definition).font(.system(size: 11, design: .monospaced)).foregroundStyle(Theme.muted).lineLimit(2)
                        },
                    ],
                    rows: rows
                )
            }
        }
        .navigationTitle("Indexes")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            res.start(load: { try await APIClient.getRows("/api/indexes") })
        }
    }
}
